pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -5513,102 +5565,149 @@ def StorageSiteWaterQualityRule(b, s, w, t):
+                b.parent_block().v_F_Trucked[p, r, t] * b.p_nu_pad[p, qc]
                 for p in b.parent_block().s_CP
                 if b.parent_block().p_CRT[p, r]
             )
-        ) == b.v_Q[r, w, t] * (
-            b.parent_block().v_F_ResidualWater[r, t]
-            + b.parent_block().v_F_TreatedWater[r, t]
+        ) == b.v_Q[r, qc, t] * b.parent_block().v_F_TreatmentFeed[r, t]
+
+        return process_constraint(constraint)
+
+    model.quality.TreatmentFeedWaterQuality = Constraint(
+        model.s_R,
+        model.s_QC,
+        model.s_T,
+        rule=TreatmentFeedWaterQualityRule,
+        doc="Treatment Feed water quality",
+    )
+
+    def TreatmentWaterQualityRule(b, r, qc, t):
+        constraint = (
+            b.v_Q[r, qc, t] * b.parent_block().v_F_TreatmentFeed[r, t]
+            == b.v_Q[r + treated_water_label, qc, t]
+            * b.parent_block().v_F_TreatedWater[r, t]
+            + b.v_Q[r + residual_water_label, qc, t]
+            * b.parent_block().v_F_ResidualWater[r, t]
         )
+
         return process_constraint(constraint)
 
     model.quality.TreatmentWaterQuality = Constraint(
         model.s_R,
-        model.s_W,
+        model.s_QC,
         model.s_T,
         rule=TreatmentWaterQualityRule,
         doc="Treatment water quality",
     )
-    # Water quality of water that has been treated
-    # NOTE: Water quality changes by treatment technologies is not currently modeled
-    def TreatedWaterWaterQualityRule(b, r, w, t):
-        constraint = (b.parent_block().v_F_TreatedWater[r, t]) == b.v_Q[
-            r + treatment_intermediate_label, w, t
-        ] * (
-            sum(
-                b.parent_block().v_F_Piped[r, p, t]
-                for p in b.parent_block().s_CP
-                if b.parent_block().p_RCA[r, p]
+
+    def TreatedWaterQualityLHSRule(b, r, wt, qc, t):
+        constraint = (
+            b.v_Q[r, qc, t]
+            * b.parent_block().v_F_TreatmentFeed[r, t]
+            * (1 - b.parent_block().p_epsilon_Treatment_Removal[r, wt, qc])
+            + b.parent_block().p_M_Flow_Conc
+            * (
+                1
+                - sum(
+                    b.parent_block().vb_y_Treatment[r, wt, j]
+                    for j in b.parent_block().s_J
+                )
             )
-            + sum(
-                b.parent_block().v_F_Piped[r, s, t]
-                for s in b.parent_block().s_S
-                if b.parent_block().p_RSA[r, s]
+            >= b.v_Q[r + treated_water_label, qc, t]
+            * b.parent_block().v_F_TreatedWater[r, t]
+        )
+
+        return process_constraint(constraint)
+
+    model.quality.TreatmentWaterQualityLHS = Constraint(
+        model.s_R,
+        model.s_WT,
+        model.s_QC,
+        model.s_T,
+        rule=TreatedWaterQualityLHSRule,
+        doc="Treatment water quality",
+    )
+
+    def TreatedWaterQualityRHSRule(b, r, wt, qc, t):
```

Comment:
A question here - in the strategic model we have constraints for ResidualWaterLHS and ResidualWaterLHS, but in the water quality we look at it from a treated water perspective instead of residual water. Is there a reason for this difference? To me, it seems that is it mathematically the same, but just curious because it's a bit confusing that it doesn't match,

Reply:
In water treatment, the removal efficiency is typically defined as the ratio of the reduction in concentration/load of the contaminant in treated water to the concentration/load of the contaminant in the feed water. This method accounts for any losses of the contaminant/water during the treatment process that are not necessarily going into the residual water.

It is therefore better to express the removal efficiency constraint in terms of treated water and, depending on whether we have material losses, write the treatment balance constraint. If there are no losses, the material balance in the current formulation of strategic model is valid otherwise it should be changed. The same rational can be applied for the treatment efficiency formulation to account for any possible water losses, depending on the technology considered.
However, I did not change the already existing strategic model part for treatment efficiency as there is already a TreatedWater rule and to minimize the changes in this PR. It is important to note that removal efficiency and treatment efficiency are independent constraints that do not need to match.

Reply:
okay, that makes sense to me. I'll resolve this conversation.